### PR TITLE
Check that the dataset has enough examples to fill a batch when creat…

### PR DESCRIPTION
**Pull request by katalinic-gc**
…ing the dataloaders in trainer

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/graphcore/trainer.py [hunk 1]
```diff
@@ -613,6 +613,30 @@ def _get_train_sampler(self) -> Optional[torch.utils.data.Sampler]:
...
+        if len(dataset) < global_batch_size:
+            mode_str = "inference_" if for_inference else ""
+            logger.warning(
+                f"The provided dataset is of length {len(dataset)}, but the total dataset batch size is {global_batch_size}. "
+                f"This batch size is calculated as:\n"
+                f"  per_device_{'eval' if for_inference else 'train'}_batch_size={micro_batch_size}\n"
+                f"* {mode_str}{replication_factor=}\n"
+                f"* {mode_str}{gradient_accumulation_steps=}\n"
+                f"* {mode_str}{device_iterations=}\n"
+                "Please disregard this warning if you believe the dataset is reporting an incorrect length, such as 1."
```
Comment (payoto) [inline optimum/graphcore/trainer.py:637]: Does this correspond to a case you've seen?

File: optimum/graphcore/trainer.py [hunk 2]
```diff
@@ -613,6 +613,30 @@ def _get_train_sampler(self) -> Optional[torch.utils.data.Sampler]:
...
+        )
+        gradient_accumulation_steps = 1 if for_inference else self.ipu_config.gradient_accumulation_steps
+        device_iterations = (
+            self.ipu_config.inference_device_iterations if for_inference else self.ipu_config.device_iterations
+        )
+        micro_batch_size = (
+            self.args.per_device_eval_batch_size if for_inference else self.args.per_device_train_batch_size
+        )
+        global_batch_size = micro_batch_size * replication_factor * gradient_accumulation_steps * device_iterations
+        if len(dataset) < global_batch_size:
```
[the remark this answers is not in the document]
Comment (katalinic-gc) [reply optimum/graphcore/trainer.py:639]: The checks are placed before the non-iterable datasets, so in theory, the dataset should have the length, but I still included the suggested change anyway.

File: optimum/graphcore/trainer.py (hunk 1 above)
Comment (katalinic-gc) [reply optimum/graphcore/trainer.py:648]: Yes, in ASR

File: optimum/graphcore/trainer.py [hunk 3]
```diff
@@ -613,6 +613,30 @@ def _get_train_sampler(self) -> Optional[torch.utils.data.Sampler]:
...
+            mode_str = "inference_" if for_inference else ""
+            logger.warning(
```
[the remark this answers is not in the document]
Comment (katalinic-gc) [reply optimum/graphcore/trainer.py:641]: Wrapped

[the remark this answers is not in the document]
Comment (katalinic-gc) [reply optimum/graphcore/trainer.py:641]: I went with a warning just because of the edge cases like length of 1 above

**Review by payoto — APPROVED** [review]
LGTM looks great thanks

**Pull request merged**